### PR TITLE
stdio/file: fix numerous issues in the fread/fwrite functions

### PR DESCRIPTION
**Pull request by ziemleszcz**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: (https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/369).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: unistd/file.c [hunk 1]
```diff
@@ -101,18 +101,26 @@ ssize_t __safe_read(int fd, void *buf, size_t size)
 			continue;
 		}
 		else {
-			break;
+			return -1;
```
Comment (Darchiv) [inline unistd/file.c:104]: > Also fixed error handling in blocking safe read to always return -1 on error.

How is this different? The `else` branch is taken when `rlen < 0`, so `if (rlen < 0)` below is still executed.

Comment (ziemleszcz) [reply unistd/file.c:104]: You're right  - disregard this comment.

File: stdio/file.c [hunk 2]
```diff
@@ -311,98 +320,186 @@ static int full_read(int fd, void *ptr, size_t size)
...
+				ret = -1;
+			}
+			else {
+				stream->bufpos = 0;
+			}
+		}
+	}
+	else {
+		if (stream->bufpos != stream->bufeof) {
+			off = lseek(stream->fd, (off_t)stream->bufpos - stream->bufeof, SEEK_CUR);
```
Comment (Darchiv) [inline stdio/file.c:366]: Should this be as-is or the other way around?

```suggestion
			off = lseek(stream->fd, (off_t)stream->bufeof - stream->bufpos, SEEK_CUR);
```

File: stdio/file.c [hunk 3]
```diff
@@ -416,67 +513,167 @@ size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream)
...
-	/* discard reading buffer */
-	if (!(stream->flags & F_WRITING)) {
-		fflush_unlocked(stream);
+	/* flush the read buffer if currently reading */
+	if ((stream->flags & F_WRITING) == 0) {
+		if (__fflush_one(stream) < 0) {
+			return 0;
+		}
+		if (stream->bufpos != stream->bufeof) {
+			/* abort if the read buffer cannot be flushed (non-seekable stream) */
```
Comment (Darchiv) [inline stdio/file.c:595]: Shouldn't `errno` be set then? (And maybe the error flag?)

File: stdio/file.c (hunk 2 above)
Comment (ziemleszcz) [reply stdio/file.c:366]: This is correct, we are seeking backwards: `-(bufeof - bufpos) = bufpos - bufeof`.

File: stdio/file.c (hunk 3 above)
Comment (ziemleszcz) [reply stdio/file.c:595]: It's implementation-specific behavior. In this case, I'm emulating glibc.

Comment (Darchiv) [reply stdio/file.c:596]: Ugh, case similar to `errno` in `malloc(0)`, so we probably can't solve this reasonably. Ack.

**Review by Darchiv — DISMISSED** [review]

**Review by Darchiv — APPROVED** [review]

**Pull request merged**